### PR TITLE
BUG: IntervalIndex.putmask with datetimelike dtypes

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Implements `NDArrayBackedExtensionArray.putmask` which will hopefully be rolled into `__array_function__` to simplify a bunch of code.

### CODE REVIEW

File: pandas/core/indexes/numeric.py [hunk 1]
```diff
@@ -122,7 +122,7 @@ def _validate_fill_value(self, value):
             # force conversion to object
             # so we don't lose the bools
             raise TypeError
-        if isinstance(value, str):
+        elif isinstance(value, str) or lib.is_complex(value):
```
Comment (jreback) [inline pandas/core/indexes/numeric.py:125]: test hitting the complex case?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/numeric.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/numeric.py:125]: yes, in tests.indexing.test_coercion

**Review by jreback — APPROVED** [review]

File: pandas/core/indexes/extension.py [hunk 2]
```diff
@@ -359,15 +359,13 @@ def insert(self, loc: int, item):
         return type(self)._simple_new(new_arr, name=self.name)
 
     def putmask(self, mask, value):
+        res_values = self._data.copy()
         try:
-            value = self._data._validate_setitem_value(value)
+            res_values.putmask(mask, value)
```
Comment (simonjayhawkins) [inline pandas/core/indexes/extension.py:364]: `self._data` is defined as `_data: ExtensionArray`

Is ExtensionIndex only for ndarray backed ExtensionArray?  or maybe will need catch AttributeError also eventually?

see also https://github.com/pandas-dev/pandas/pull/35032#issuecomment-695218091 (sort of related).

Comment (simonjayhawkins) [reply pandas/core/indexes/extension.py:364]: Ah my bad for not checking the code thoroughly before commenting and just looking at the diff.

I wasn't aware that you recently added a NDArrayBackedExtensionIndex.

Doesn't this add more complexity? what is the motivation for this?

Comment (jbrockmendel) [reply pandas/core/indexes/extension.py:364]: > Doesn't this add more complexity? what is the motivation for this?

Its just taking methods that are duplicated in CategoricalIndex+DatetimeIndexOpsMixin and putting them in one place.  I guess you could say that "adds complexity" bc there is an additional step in the inheritance chain?

**Pull request merged**